File: Dynamic Programming/Longest_Increasing_Subsequence.cpp
// Longest Increasing Subsequence - DP Approach
// Author - rudrajit1729

#include<bits/stdc++.h>
using namespace std;

void generate(int A[], int n, int max = 20)
{
    // Generates a list of random numbers
    for(int i = 0; i < n; i++)
    {
        A[i] = rand()%max + 1;
    }
}

int solve(int A[], int n)
{
    // returns the length of the longest subsequence
    int DP[n]; // Memo table to store length of the subsequences for the indices
    // Building table bottom up
    for(int i = n-1; i>=0; i--)
    {
        vector<int>choices;
        choices.push_back(1); // Single element always a subsequence - len = 1
        for(int j = i+1; j < n; j++)
        {
            // Iterating through the elements to the right if we find a element bigger than given element
            // We include that element to list of choices along with the length of subsequence at that position
            if (A[j] > A[i]) choices.push_back(DP[j] + 1);
        }
        DP[i] = *max_element(choices.begin(), choices.end());
    }
    return *max_element(DP, DP + n);
}

void printsol(vector<int> ans, int n)
{
    // Utility to print the solution
    cout<<" (";
    for(int i = n-1; i > 0; i--)
    cout<<ans[i]<<", ";
    cout<<ans[0]<<")\n";
}

int answer(int A[], int n)
{
    // DP[i] - Longest increasing subsequence ending with A[i]
    vector<vector<int>> DP(n);
    // Iterate from index n-1
    for(int i = n-1; i >=0; i--)
    {
        // Traverse through j greater than i
        for(int j = i + 1; j < n; j++)
        {
            //  DP[i] = max(DP[j]) + A[i]
            //  j < i and A[j] < A[i]
            if(A[i] < A[j] && (DP[i].size() < DP[j].size() + 1))
                DP[i] = DP[j];
        }
        // DP[i] starts with A[i]
        DP[i].push_back(A[i]);
    }
    // Maximum subsequence now printed. DP[i] now stores increasing subsequence 
    vector<int> max = DP[0];
    for(vector<int> x:DP)
    {
        if (x.size() > max.size())
            max = x;
    }
    cout<<"Subsequence :";
    printsol(max, max.size());
    return max.size();
}

int main()
{
    srand(time(NULL));
    int n;
    cout << "Enter number of elements : ";
    cin >> n;
    int A[n];
    // generate(A, n);
    cout<<"Enter elements : ";
    for(int i = 0; i < n; i++)
        cin >> A[i];

    // cout<<"Longest Common Subsequence"<< solve(A, n);
    cout<<"Longest Common Subsequence : "<<endl;
    cout<<"Length of Subsequence : "<<answer(A, n);
}

File: Dynamic Programming/Longest_Increasing_Subsequence.py
# Longest Increasing Subsequence - DP Approach
# Author - rudrajit1729

from random import randint

def generate(n, max = 20):
    ''' Generates a list of random numbers'''
    l = []
    for _ in range(n):
        l.append(randint(1, max))
    return l

def solve(A):
    ''' returns the length of the longest subsequence'''
    n = len(A)
    DP = [] * n # Memo table to store length of the subsequences for the indices
    # Building table bottom up
    for i in range(n-1, -1, -1):
        choices = [1] # Single element always a subsequence - len = 1
        for j in range(i+1, n):
            # Iterating through the elements to the right if we find a element bigger than given element
            # We include that element to list of choices along with the length of subsequence at that position
            if A[j]>A[i]:
                choices.append(DP[j] + 1)
        DP[i] = max(choices)
    return max(DP)

def answer(A):
    n = len(A)
    # DP[i] - Longest increasing subsequence starting with A[i]
    DP = [[] for i in range (n)]

    # Iterate from index n-1
    for i in range(n-1, -1, -1):
        # Traverse through j greater than i
        for j in range(i+1, n):
            # DP[i] = max(DP[j]) + A[i]
            # j < i and A[j] < A[i]
            if A[i] < A[j] and (len(DP[i]) < len(DP[j]) + 1):
                DP[i] = DP[j].copy()
        # DP[i] starts with A[i]
        DP[i].insert(0, A[i])
        

    # DP[i] now stores increasing subsequence 
    maxx = DP[0]
    for seq in DP:
        if len(seq) > len(maxx):
            maxx = seq

    print("Subsequence : ", end = "")
    print(maxx) 
    return len(maxx) 

def main():
    A = list(map(int, input("Enter elements : ").split()))
    #A = generate(7, 10)
    print(A)
    print("Longest Common Subsequence : ")
    #print(solve(A))
    print("Length = ", answer(A))

if __name__ == "__main__":
    main()

   


